Add client hint docs: Downlink, ECT, RTT

hamishwillee opened the pull request:
This header is in the Network Information spec here: https://wicg.github.io/netinfo/#rtt-request-header-field

There is not all that much information, but you can see it is associated with the [NetworkInformation.rtt](https://developer.mozilla.org/en-US/docs/Web/API/NetworkInformation/rtt) and there is some useful general guidance in https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/client-hints.

Spec update is in https://github.com/mdn/browser-compat-data/pull/10951

This is another part of #1408

hamishwillee committed:
RTT Client hint header docs

hamishwillee commented on files/en-us/web/http/headers/rtt/index.html:
```diff
@@ -0,0 +1,69 @@
+  </tbody>
+ </table>
+
+ <p>The hint allows a server to choose what information is sent based on the network responsiveness/latency. For example, it might choose not to respond to some requests.</p>
```
Note, I am not completely happy with this. It is true that this is something it might do, but I'm not sure what the limits are, if any, on what a server can do. For example, could it decide to respond with a low res image instead?

Elchi3 replied:
I'm not an expert with client hints. I thought upon a bad connection `RTT` is useful to serve **less** resources and `Downlink` is useful to serve **smaller** resources?

hamishwillee replied:
@Elchi3 Thanks.  FYI only [the example chrome gives](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/client-hints#helping_users_on_slow_networks) use the hints to determine a "network score" that then decides whether to serve web fonts, hero images, carousel and associated javascript. In other words "kind of" :-). 

Upshot though is that my question is also kind of answered - the server can do what it likes.

I still don't really get the RTT case though. Server requests RTT, client provides it, so now server knows that latency is low and decides not to send web fonts, and a carousel. Presumably the request for these is still made and this implies that the page has to configure itself to reasonably work without that information.

hamishwillee committed:
Add pages for all the network client hints

hamishwillee committed:
Cross link the WebAPI versions to the Headers

hamishwillee committed:
Restructure main headers page to make it extensible for new types

hamishwillee committed:
Better cross linking on Accept-CH

hamishwillee committed:
Restructure client hints glossary

hamishwillee committed:
Merge branch 'main' into http_rtt

hamishwillee commented:
@Elchi3 OK, this got bigger
- I added the missing network headers ECT and Downlink
- Updated Save-Data to better cross link.
- Cross linked all of these to the equivalent web API 
- Added sections for client headers in both the main headers page and in the client headers page. 

I think this is a good place to stop for now. In future though it might be worth having a client headers guide much like https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/client-hints and cut back the glossary entry. This gives us a structure to work with now though.

Can you please review again :-)

Elchi3 commented on files/en-us/web/http/headers/ect/index.html:
```diff
@@ -0,0 +1,77 @@
+
+<h2 id="Syntax">Syntax</h2>
+
+<pre class="brush: http">ect: &lt;value&gt;</pre>
```
```suggestion
<pre class="brush: http">ECT: &lt;value&gt;</pre>
```

Elchi3 replied:
Describe \<value> in a Directives section?

Elchi3 commented on files/en-us/web/http/headers/ect/index.html:
```diff
@@ -0,0 +1,77 @@
+
+<h2 id="Examples">Examples</h2>
+
+<p>Server first needs to opt in to receive the <code>Downlink</code> header by sending the response headers {{HTTPHeader("Accept-CH")}} containing <code>Downlink</code>.</p>
```
```suggestion
<p>A server first needs to opt in to receive the <code>ECT</code> header by sending the {{HTTPHeader("Accept-CH")}} response header containing <code>ECT</code>.</p>
```

Elchi3 commented on files/en-us/web/http/headers/ect/index.html:
```diff
@@ -0,0 +1,77 @@
+
+<pre class="brush: http">Accept-CH: ECT</pre>
+
+<p>Then on subsequent requests the client might send <code>Downlink</code> header back:</p>
```
```suggestion
<p>Then on subsequent requests the client might send an <code>ECT</code> header back:</p>
```

Elchi3 commented on files/en-us/web/http/headers/ect/index.html:
```diff
@@ -0,0 +1,77 @@
+  <ul>
+    <li>{{HTTPHeader("Downlink")}}</li>
+    <li>{{HTTPHeader("RTT")}}</li>
+    <li>{{HTTPHeader("ECT")}}</li>
```
```suggestion
```

Elchi3 commented on files/en-us/web/http/headers/downlink/index.html:
```diff
@@ -0,0 +1,76 @@
+  <li><a href="https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/client-hints">Adapting to Users with Client Hints</a> (developer.google.com)</li>
+ <li>Network client hints
+  <ul>
+    <li>{{HTTPHeader("Downlink")}}</li>
```
```suggestion
```

Elchi3 commented on files/en-us/web/http/headers/index.html:
```diff
@@ -76,42 +76,64 @@ <h2 id="Caching">Caching</h2>
 
-<p>HTTP {{Glossary("Client_hints", "Client hints")}} are a set of request headers that provide useful information about the client such as device type and network conditions, and allow the server to optimize what it serves for those conditions
-  (<a href="https://datatracker.ietf.org/doc/html/rfc8942"><cite>[RFC8942]</cite></a>).</p>
+<p>HTTP {{Glossary("Client_hints", "Client hints")}} are a set of request headers that provide useful information about the client such as device type and network conditions, and allow servers to optimize what is serves for those conditions.</p>
```
```suggestion
<p>HTTP {{Glossary("Client_hints", "Client hints")}} are a set of request headers that provide useful information about the client such as device type and network conditions, and allow servers to optimize what is served for those conditions.</p>
```

Elchi3 commented on files/en-us/web/http/headers/index.html:
```diff
@@ -76,42 +76,64 @@ <h2 id="Caching">Caching</h2>
+  <dd>Client device pixel ratio (DPR), which is the number of physical device pixels corresponding to every CSS pixel.</dd>
+  <dt>{{HTTPHeader("Viewport-Width")}} {{experimental_inline}}</dt>
+  <dd>
+  <div>
```
I think you don't need the div and p elements in dd.

hamishwillee replied:
Removed. Note the  `p` was there previously because there are two paragraphs - but I have decided to merge these into one. Thanks.

Elchi3 commented on files/en-us/web/http/headers/rtt/index.html:
```diff
@@ -0,0 +1,77 @@
+  </tbody>
+ </table>
+
+<p>The RTT value is rounded to the nearest 25 milliseconds to prevent fingerprinting; There are many other mechanisms an attacker might use to obtain similar round-trip information.</p>
```
Same here, I would move this into a "Directives" section providing details about \<number>

Elchi3 commented on files/en-us/web/http/headers/rtt/index.html:
```diff
@@ -0,0 +1,77 @@
+
+<h2 id="Examples">Examples</h2>
+
+<p>Server first needs to opt in to receive the <code>RTT</code> header by sending the response headers {{HTTPHeader("Accept-CH")}} containing <code>RTT</code>.</p>
```
```suggestion
<p>A server first needs to opt in to receive the <code>RTT</code> header by sending the {{HTTPHeader("Accept-CH")}} response header containing <code>RTT</code>.</p>
```

Elchi3 commented on files/en-us/web/http/headers/rtt/index.html:
```diff
@@ -0,0 +1,77 @@
+
+<pre class="brush: http">Accept-CH: RTT</pre>
+
+<p>Then on subsequent requests the client might send <code>RTT</code> header back:</p>
```
```suggestion
<p>Then on subsequent requests the client might send an <code>RTT</code> header back:</p>
```

Elchi3 commented on files/en-us/web/http/headers/rtt/index.html:
```diff
@@ -0,0 +1,77 @@
+ <li>Network client hints
+  <ul>
+    <li>{{HTTPHeader("Downlink")}}</li>
+    <li>{{HTTPHeader("RTT")}}</li>
```
```suggestion
```

hamishwillee replied:
@Elchi3 FYI the inclusion was deliberate. Now that Yari does the right thing and renders this "link to self" as a black unclickable link, it makes this list into a clear "set", which I quite like (and is easy to copy-paste when documenting new mthods) :-). Upshot I will do as you have suggest but I don think of this as broken.

Elchi3 commented on files/en-us/web/http/headers/save-data/index.html:
```diff
@@ -112,14 +114,18 @@ <h2 id="Browser_compatibility">Browser compatibility</h2>
+       <li>{{HTTPHeader("Downlink")}}</li>
+       <li>{{HTTPHeader("RTT")}}</li>
+       <li>{{HTTPHeader("ECT")}}</li>
+       <li>{{HTTPHeader("Save-Data")}}</li>
```
```suggestion
```

Elchi3 reviewed (changes requested):
Thanks for your work! Some comments.

hamishwillee commented on files/en-us/glossary/client_hints/index.html:
```diff
@@ -8,17 +8,18 @@
 
-<p>A server must announce that it supports client hints, using the {{HTTPHeader("Accept-CH")}} header to specify the hints that it is interested in recieving.
-  When a client that supports client hints receives the <code>Accept-CH</code> header it appends client hint headers that match the advertised field-values to subsequent requests.</p>
+<p>The set of "hint" headers are listed in the topic <a href="/en-US/docs/Web/HTTP/Headers#client_hints">HTTP Headers</a> and <a href="#hint_types">summarised below</a>.</p>
```
Arrgg. Thanks. US-English is my bane.

hamishwillee committed:
Update files/en-us/glossary/client_hints/index.html

Co-authored-by: Florian Scholz <fs@florianscholz.com>

hamishwillee committed:
fix typo glossary/client_hints

Co-authored-by: Florian Scholz <fs@florianscholz.com>

hamishwillee committed:
Typo: missing fullstop in glossary/client_hints

Co-authored-by: Florian Scholz <fs@florianscholz.com>

hamishwillee committed:
Typo: missing fullstop

Co-authored-by: Florian Scholz <fs@florianscholz.com>

hamishwillee committed:
Revert Specifications macro in Accept-CH

hamishwillee commented on files/en-us/web/http/headers/accept-ch/index.html:
```diff
@@ -51,20 +57,7 @@ <h2 id="Examples">Examples</h2>
-    </tr>
-  </tbody>
-</table>
+{{Specifications}}
```
Reverted. Can't wait for this to work.

hamishwillee committed:
Update files/en-us/web/http/headers/downlink/index.html

Co-authored-by: Florian Scholz <fs@florianscholz.com>

hamishwillee committed:
Update files/en-us/web/http/headers/downlink/index.html

Co-authored-by: Florian Scholz <fs@florianscholz.com>

hamishwillee committed:
Update files/en-us/web/http/headers/ect/index.html

Co-authored-by: Florian Scholz <fs@florianscholz.com>

hamishwillee committed:
Update files/en-us/web/http/headers/rtt/index.html

Co-authored-by: Florian Scholz <fs@florianscholz.com>

hamishwillee committed:
Update files/en-us/web/http/headers/index.html

Co-authored-by: Florian Scholz <fs@florianscholz.com>

hamishwillee committed:
Update files/en-us/web/http/headers/rtt/index.html

Co-authored-by: Florian Scholz <fs@florianscholz.com>

hamishwillee committed:
Update files/en-us/web/http/headers/rtt/index.html

Co-authored-by: Florian Scholz <fs@florianscholz.com>

hamishwillee committed:
Update files/en-us/web/http/headers/ect/index.html

Co-authored-by: Florian Scholz <fs@florianscholz.com>

hamishwillee committed:
Update files/en-us/web/http/headers/save-data/index.html

Co-authored-by: Florian Scholz <fs@florianscholz.com>

hamishwillee committed:
Update files/en-us/web/http/headers/ect/index.html

Co-authored-by: Florian Scholz <fs@florianscholz.com>

hamishwillee committed:
Update files/en-us/web/http/headers/ect/index.html

Co-authored-by: Florian Scholz <fs@florianscholz.com>

hamishwillee committed:
Update files/en-us/web/http/headers/downlink/index.html

Co-authored-by: Florian Scholz <fs@florianscholz.com>

hamishwillee committed:
Update files/en-us/web/http/headers/downlink/index.html

Co-authored-by: Florian Scholz <fs@florianscholz.com>

hamishwillee commented on files/en-us/web/http/headers/downlink/index.html:
```diff
@@ -0,0 +1,76 @@
+  </tbody>
+ </table>
+
+<p>The Downlink value is rounded to the nearest 25 kilobits per second to prevent fingerprinting; There are many other mechanisms an attacker might use to obtain similar information.</p>
```
@Elchi3 Thanks. Fixed for the three cases.

I copied DPR header which omits this rather than https://developer.mozilla.org/en-US/docs/MDN/Structures/Page_types/HTTP_header_page_template (which is correct). I will fix up the DPR case when I do the device client hints updates next.

hamishwillee committed:
Fixes following feedback from Florian (directives sections)

hamishwillee commented:
@Elchi3 Thanks for the detailed review - I seem to have injected more errors than usual :-(. 

I accepted all changes, added directives sections, reverted the specifications section. Note that I kept the specifications macro for the **new** header docs. Yes they show missing info now, but I'm hoping that the spec issue will be resolved soon and don't think it worth adding the manual table while we wait. Happy to add a manual table if you think this is a "requirement".

Elchi3 commented:
> Note that I kept the specifications macro for the **new** header docs. Yes they show missing info now, but I'm hoping that the spec issue will be resolved soon and don't think it worth adding the manual table while we wait. Happy to add a manual table if you think this is a "requirement".

This was the correct call! These headers use https://wicg.github.io/netinfo/ which does in fact work as soon as the latest BCD is on MDN production in a few hours. What doesn't work right now is HTTP features that reference IETF specs. You can follow https://github.com/w3c/browser-specs/issues/280 to be notified when it's available upstream.

Elchi3 reviewed (approved):
Thanks Hamish, fantastic work! 👍
